scvi/external/cell2location/autoguide.py

```diff
@@ -0,0 +1,384 @@
+from contextlib import ExitStack  # python 3
+
+import numpy as np
+import pyro
+import pyro.distributions as dist
+import torch
+from pyro.distributions import constraints
+from pyro.distributions.transforms import SoftplusTransform
+from pyro.distributions.util import sum_rightmost
+from pyro.infer.autoguide import AutoGuide
+from pyro.infer.autoguide import AutoGuideList as PyroAutoGuideList
+from pyro.infer.autoguide.guides import _deep_getattr, _deep_setattr
+from pyro.infer.autoguide.utils import helpful_support_errors
+from pyro.nn import PyroModule, PyroParam
+from torch.distributions import biject_to, transform_to
+
+from scvi.nn import FCLayers
+
+
+class FCLayersPyro(FCLayers, PyroModule):
+    pass
+
+
+class AutoGuideList(PyroAutoGuideList):
+    def quantiles(self, quantiles, *args, **kwargs):
+        """
+        Returns the posterior quantile values of each latent variable.
+
+        Parameters
+        ----------
+        quantiles
+            A list of requested quantiles between 0 and 1.
+
+        Returns
+        -------
+        A dict mapping sample site name to quantiles tensor.
+        """
+
+        result = {}
+        for part in self:
+            result.update(part.quantiles(quantiles, *args, **kwargs))
+        return result
+
+
+@biject_to.register(constraints.positive)
+@transform_to.register(constraints.positive)
+def _transform_to_positive(constraint):
+    return SoftplusTransform()
+
+
+class AutoNormalEncoder(AutoGuide):
```

Comment:
Hi @vitkl, @martinjankowiak and @eb8680 and I have been wanting something like this in Pyro for a long time, and we'd be happy to maintain part or all of this in the Pyro codebase (with tests, docs, etc.).

Reply:
Just created a PR here: https://github.com/pyro-ppl/pyro/pull/2849